scipy/optimize/tnc/_tnc_module.pyx

```diff
@@ -0,0 +1,204 @@
+# cython: language_level=3, boundscheck=False
+
+from libc.string cimport memcpy
+import numpy as np
+cimport numpy as np
+cimport cython
+
+
+np.import_array()
+
+ctypedef np.float64_t float64_t
+
+
+cdef extern from "tnc.h":
+    ctypedef void tnc_callback(double[], void*) except *
+    ctypedef int tnc_function(double[], double*, double[], void*) except 1
+
+    cdef enum tnc_rc:
+        TNC_MINRC
+        TNC_ENOMEM
+        TNC_EINVAL
+        TNC_INFEASIBLE
+        TNC_LOCALMINIMUM
+        TNC_FCONVERGED
+        TNC_XCONVERGED
+        TNC_MAXFUN
+        TNC_LSFAIL
+        TNC_CONSTANT
+        TNC_NOPROGRESS
+        TNC_USERABORT
+
+    cdef int tnc(int n, double x[], double *f, double g[],
+            tnc_function *function, void *state,
+            double low[], double up[], double scale[], double offset[],
+            int messages, int maxCGit, int maxnfeval, double eta, double stepmx,
+            double accuracy, double fmin, double ftol, double xtol, double pgtol,
+            double rescale, int *nfeval, int *niter, tnc_callback *tnc_callback) except *
+
+
+cdef struct s_pytnc_state:
+  void *py_function
+  void *py_callback
+  int n
+  int failed
+ctypedef s_pytnc_state pytnc_state
+
+
+cdef int function(double x[], double *f, double g[], void *state) except 1:
+    # return 0 if no error, 1 if error
+    # However, if the user function raises Exceptions Cython should deal
+    # with it.
+
+    cdef:
+        pytnc_state *py_state
+        int n
+        double *x_data
+        double *g_data
+
+    py_state = <pytnc_state *>state
```

Comment:
So I'm proposing the `py_state.failed` check should go here.

Reply:
I added that.